File: Plugins/Orimath.Basics.View/ViewModels/AttachedObservableCollection.cs
using System;
using System.Linq;
using Mvvm;
using Orimath.Plugins;
using ApplicativeProperty;

namespace Orimath.Basics.View.ViewModels
{
    public sealed class AttachedObservableCollection<TModel, TViewModel> : ResettableObservableCollection<TViewModel>, IDisposable
    {
        private readonly IDisposable _disposer;
        private readonly IDispatcher _dispatcher;
        private readonly Func<TModel, TViewModel> _mapper;
        private readonly Action<TViewModel> _onRemove;

        public AttachedObservableCollection(
            IDispatcher dispatcher,
            IReactiveCollection<TModel> source,
            Func<TModel, TViewModel> mapper,
            Action<TViewModel> onRemove)
            : base(source.Select(mapper))
        {
            _dispatcher = dispatcher;
            _mapper = mapper;
            _onRemove = onRemove;
            _disposer = source.Subscribe(Source_CollectionChanged);
        }

        private async void Source_CollectionChanged(CollectionChange<TModel> e)
        {
            await _dispatcher.SwitchToUI();
            switch (e)
            {
                case CollectionChange<TModel>.Add add:
                    foreach (var item in add.items) Add(_mapper(item));
                    break;

                case CollectionChange<TModel>.Remove remove:
                    var length = Count;
                    for (var index = length - 1; index >= length - remove.items.Count; index--)
                    {
                        _onRemove(this[index]);
                        RemoveAt(index);
                    }
                    break;

                case CollectionChange<TModel>.Replace replace:
                    _onRemove(this[replace.index]);
                    this[replace.index] = _mapper(replace.newItem);
                    break;

                case CollectionChange<TModel>.Reset reset:
                    foreach (var item in this) _onRemove(item);
                    Reset(reset.newItems.Select(_mapper));
                    break;
            }
        }

        public void Dispose()
        {
            _disposer.Dispose();
        }
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/AttachedObservableCollection.fs
namespace Orimath.Basics.View.ViewModels
open System
open Orimath.Controls
open Orimath.Plugins
open ApplicativeProperty

type AttachedObservableCollection<'Model, 'ViewModel>
    (dispatcher: IDispatcher,
     source: IReactiveCollection<'Model>,
     mapping: 'Model -> 'ViewModel,
     onRemove: 'ViewModel -> unit) as this =
    inherit ResettableObservableCollection<'ViewModel>(source |> Seq.map mapping)
    let disconnector =
        source |> Observable.subscribe2(this.SourceCollectionChanged)
    
    member private this.SourceCollectionChanged(e) =
        dispatcher.UI.Invoke(fun () ->
            match e with
            | CollectionChange.Add(index, items) ->
                if index = this.Count then
                    for item in items do this.Add(mapping item)
                else
                    for i = 0 to items.Count - 1 do
                        this.Insert(index + i, mapping items[i])
            | CollectionChange.Remove(index, items) ->
                let last = index + items.Count
                for index = last - 1 downto index do
                    onRemove(this[index])
                    this.RemoveAt(index)
            | CollectionChange.Replace(index, _, newItem) ->
                onRemove(this[index])
                this[index] <- mapping newItem
            | CollectionChange.Reset(_, newItems) ->
                this |> Seq.iter onRemove
                this.Reset(newItems |> Seq.map mapping))

    member _.Dispose() = disconnector.Dispose()

    interface IDisposable with
        member _.Dispose() = disconnector.Dispose()
